rfd/0163-vnet.md

```diff
@@ -0,0 +1,996 @@
+routing table, and configure VNet as a DNS nameserver.
+
+To create the TUN interface we will import `golang.zx2c4.com/wireguard/tun`
+which provides cross-platform methods for creating TUN interfaces and reading
+from and writing to them.
+It supports MacOS, Windows, and Linux.
+
+The following applies to MacOS only, we will have other platform-specific
+methods when we add support for Windows and Linux.
+
+To assign an IP address to the interface, we will use `os/exec` to run the
+`ifconfig` command:
+
+```shell
+ifconfig utun4 100.64.0.1 100.64.0.1 up
+```
+
+To assign a IP routes to use the TUN interface, we will use `os/exec` to run the
+`route` command:
+
+```shell
+route add -net 100.64.0.0/10 -interface utun4
+route add -inet6 fdec:146c:95f6::/64 -interface utun4
+```
+
+To configure split DNS, we will add the VNet DNS server to files under
+`/etc/resolver` for each cluster the user is logged in to and each custom DNS
+zone.
+This will be updated each time the user logs in to a new cluster while VNet is
+enabled.
+
+```shell
+$ # cluster with proxy at teleport.example.com and custom DNS zone example.com
+$ ls /etc/resolver/
+teleport.example.com example.com
+$ cat /etc/resolver/teleport.example.com
+nameserver 100.127.100.127
+$ cat /etc/resolver/example.com
+nameserver 100.127.100.127
+```
+
+The first version of VNet will only support MacOS so this is written with that
+in mind, but we can implement this in a similar way for Windows and Linux.
+
+#### Daemon
+
+All system configuration must run as a privileged user (root). To achieve that, VNet starts a daemon
+with `tsh vnet daemon` that runs as root, a Launch Daemon in macOS terms. The daemon handles
+creating a TUN device (and passing back its [file descriptor over a gRPC
+connection on Unix-like systems](https://pkg.go.dev/github.com/edwarnicke/grpcfd) or [GUID on
+Windows](https://github.com/WireGuard/wireguard-go/blob/12269c2761734b15625017d8565745096325392f/tun/tun_windows.go#L66-L68))
+as well as DNS configuration.
+
+The main reason behind adding a daemon is that otherwise we'd be forced to prompt for a password
+each time we need to carry out privileged operations.
+
+On macOS, the daemon is added through [`SMAppService`](https://developer.apple.com/documentation/servicemanagement/smappservice?language=objc)
+and a plist that lives inside the app bundle. This has two nice properties. First, the daemon and
+the plist are signed by us, preventing 3rd parties from making any changes. Second, since the daemon
+plist lives in the app bundle, it is automatically removed when the user removes tsh.app or Teleport
+Connect.app.
+
+`SMAppService` requires macOS 13.0+. Since it only works with signed packages, in development we can
+substitute it with an osascript wrapper that runs the gRPC server as root and shuts down with the
+parent process.
+
+The daemon allows other processes to talk to it through a gRPC service available over a Unix socket
```

Comment:
What authentication will we use here?  Can you document what this API to the daemon will be?

Reply:
As for the API, it's in [the Proto Specification section](https://github.com/gravitational/teleport/blob/rfd/0163-vnet/rfd/0163-vnet.md#daemon-service), sorry for not mentioning that!

I didn't plan on using any gRPC authentication because I couldn't come up with any method that would actually offer any security here. The gRPC service is used purely for IPC and we want an unprivileged client to talk to a privileged process.

I'd appreciate any ideas here as I don't have much experience with exposing access to a privileged process like this.

I saw that with XPC, it's possible to limit access to a daemon through codesigned entitlements, i.e. only programs codesigned by us can talk to the daemon. But XPC is macOS-only and we need to eventually port VNet to Windows and Linux.

Reply:
Nic suggested changing the owner of the socket file to the desired user and then setting permissions to 660. Doing so before the connection is established is tricky as registering a daemon boils down to telling macOS which plist to register and the plist needs to be embedded within the app – we can't generate it dynamically to insert uid/username there.

Once a client starts a session, we can reject calls from other clients with a different pid and uid. But that ultimately doesn't address the fact, which I guess might be the biggest issue here, that any process can talk to that privileged daemon.

Reply:
Just brainstorming, but what about a multi-tenant system?  For example should this feature work on a shared host which has multiple users connecting to it?